core/src/main/java/org/elasticsearch/action/get/MultiGetResponse.java

```diff
@@ -149,6 +186,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public static MultiGetResponse fromXContent(XContentParser parser) {
+        List<MultiGetItemResponse> items = new ArrayList<>();
+        try {
+            String currentFieldName = null;
+            for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+                switch (token) {
+                    case FIELD_NAME:
+                        currentFieldName = parser.currentName();
+                        break;
+                    case START_ARRAY:
+                        if (Fields.DOCS.equals(currentFieldName) == false) {
+                            throw new ElasticsearchParseException("Unexpected field [" + currentFieldName + "]");
+                        }
+                        for (Token innerToken = parser.nextToken(); innerToken != Token.END_ARRAY; innerToken = parser.nextToken()) {
+                            if (innerToken == Token.START_OBJECT) {
+
+                                // Parsing this in a streaming manner means that at least up to 4 fields need be
+                                // keep tracked of here and then GetResult.fromXContent(...) needs to be broken up.
+                                // This is makes the parsing code complex, whereas reading each json object into a map and
+                                // then creating a parser from that is a bit inefficient, but makes the parsing code less
+                                // complex. I think this is the right trade off. Unless we introduce a new xcontent format
+                                // in mget response for the hl rest client that is enabled via a parameter that renders
+                                // an unambiguous format.
```

Comment:
It makes me a bit sad to have to go with maps here, yet I see why it is hard to do it differently.
@cbuescher any idea on making this work without maps? Here is an example response for clarify:

```
{
  "docs" : [
    {
      "_index" : "my_index",
      "_type" : "my_type",
      "_id" : "1",
      "_version" : 1,
      "found" : true,
      "_source" : {
        "user" : [
          {
            "field1" : 1,
            "field2" : 2
          },
          {
            "field1" : 3,
            "field2" : 4
          }
        ]
      }
    },
    {
      "_index" : "test",
      "_type" : "type",
      "_id" : "2",
      "error" : {
        "root_cause" : [
          {
            "type" : "index_not_found_exception",
            "reason" : "no such index",
            "resource.type" : "index_expression",
            "resource.id" : "test",
            "index_uuid" : "_na_",
            "index" : "test"
          }
        ],
        "type" : "index_not_found_exception",
        "reason" : "no such index",
        "resource.type" : "index_expression",
        "resource.id" : "test",
        "index_uuid" : "_na_",
        "index" : "test"
      }
    }
  ]
}
```

Reply:
As far as I understand after a first glace the difference here is that the objects in `docs` can eithe be failures or sth. that can already be parsed by GetResult#fromXContent(). The only thing the failure items have is an `error` field. Ideally we could reuse the GetResult parser so it optionally accepts an `error` field. Unfortunaltely the current GetResult parser isn't easily extendable, but could we add a flag to it that allows optionally parsing the `error` field and have GetResult contain an optional Failure? 
Another option would be to rewrite GetResult#fromXContent to use Object parser, then we could have a common "declareFields(Parser ...)" method that sets up the parser with the parts that GetResult needs and reuse that setup method plus a parser for the "error" field  for a specialized MultiGetResult or something.
I think that way we could avoid parsing to a generic map here, which I would also appreciate.

Reply:
@javanna @cbuescher I'll look into changing the `GetResult#fromXContent(...)` method, so that we don't have to parse into a generic map here.